crates/kubelet/src/pod/queue.rs

```diff
@@ -19,79 +30,166 @@ use crate::provider::Provider;
+                                Ok(()) => info!("Pod {} state machine exited without error", name),
+                                Err(e) => error!("Pod {} state machine exited with error: {:?}", name, e),
+                            }
+                        }
+                    }
+
+                    info!("Pod {} waiting for deregistration.", name);
+                    loop {
+                        tokio::time::delay_for(std::time::Duration::from_secs(1)).await;
+                        if *check_pod_deleted.read().await {
+                            info!("Pod {} deleted.", name);
+                            break;
+                        }
+                    }
+                    pod_state.async_drop().await;
+                    drop(pod_state);
+
+                    let pod_client: kube::Api<KubePod> =
+                        kube::Api::namespaced(task_client, pod.namespace());
+                    let dp = kube::api::DeleteParams {
+                        grace_period_seconds: Some(0),
+                        ..Default::default()
+                    };
+                    match pod_client.delete(&pod.name(), &dp).await {
+                        Ok(_) => {
+                            info!("Pod {} deregistered.", name);
+                        }
+                        Err(e) => {
+                            error!("Unable to deregister {} with Kubernetes API: {:?}", name, e);
+                        }
+                    }
+                });
+            }
+            _ => anyhow::bail!("Pod with initial event not Added: {:?}", &initial_event),
+        }
+
+        tokio::spawn(async move {
             while let Some(event) = receiver.recv().await {
                 // Watch errors are handled before an event ever gets here, so it should always have
                 // a pod
-                let pod = pod_from_event(&event).unwrap();
-                if let Err(e) = provider.handle_event(event).await {
-                    if let Err(e) = error_sender.send((pod, e)).await {
-                        error!("Unable to send error to status updater: {:?}", e)
+                match event {
+                    WatchEvent::Modified(pod) => {
+                        info!("Pod {} modified.", Pod::new(pod.clone()).name());
+                        // Not really using this right now but will be useful for detecting changes.
+                        let pod = Pod::new(pod);
+                        // TODO, detect other changes we want to support, or should this just forward the new pod def to state machine?
+                        if let Some(_timestamp) = pod.deletion_timestamp() {
+                            *(pod_deleted.write().await) = true;
+                        }
+                    }
+                    WatchEvent::Deleted(pod) => {
+                        // I'm not sure if this matters, we get notified of pod deletion with a
```

Comment:
I think the proper thing here would be to do `WatchEvent::Modified(pod) | WatchEvent::Deleted(pod)` for the block above if we want to run our deregister logic anyway. Right now, if someone force deletes the pod, it wouldn't run the cleanup on our end. The only thing is that error handling would have to manage if the pod is already deleted and ignore it instead of logging the error

Reply:
Ah, good point.